Adding support for building custom editor displays.

rcarcasses opened the pull request:
Greetings awesome team! I have been working recently in the wordpress plugin to expose the h5p editor not only through the wordpress admin but also from a custom url. For example, with the modifications I have added, one can show the editorh5p-editor by doing the following:

        
      {
               //  ... somewhere in the constructor of a plugin

               // we want to show a clean editor if requested, here we modify the template
		add_filter( 'page_template', array($this, 'h5p_editor_page_template'));
               // in order to actually save the h5p created in the custom page, we need something like this
		add_action('init', array($this, 'process_new_content'));
	} // End __construct ()

	function process_new_content() {
		// H5P_EDITOR_SLUG is a static constant = 'h5p-editor'..., the custom url
		if ($_SERVER['REQUEST_METHOD'] === 'POST' &&
			$_SERVER['REQUEST_URI'] == '/' . self::H5P_EDITOR_SLUG &&
	        isset($_POST['action']) && $_POST['action'] == 'create')	{
			//if everything is fine, 'good' will be returned, useful if the editor is embedded inside
                        // a more complex javascript application...
			$this->h5p_admin->process_new_content('good');
		}
	}

        // this customize the display of the specific page where the custom h5p editor will be displayed
	function h5p_editor_page_template($page_template) {
		if (is_page(self::H5P_EDITOR_SLUG)) {
			$page_template = dirname( __FILE__  ) . '/tpl-h5p-editor.php';
		}

		return $page_template;
	}

the content of tpl-h5p-editor.php is:
```
<?php

	wp_head();
	require_once WP_PLUGIN_DIR . '/h5p/autoloader.php';

	$plugin_admin = H5P_Plugin_Admin::get_instance();
	$plugin_admin->display_new_content_page(dirname(__FILE__) . '/min-h5p-editor.php');
```
the content of min-h5p-editor.php would be a custom version of the file admin/views/new-content.php, for instance.

Basically this pull request ask for the creation of two public functions in the H5P_Plugin_Admin and a slightly modification of the process_new_content method in the H5PContentAdmin to avoid redirecting always to the admin page.

If you like this, please accept it and thank you for your nice work!

icc commented:
This is a good idea, the code should be flexible and reusable by other plugins. May I ask for which purpos you're using this? 

Your code looks OK and it doesn't involve too many changes. This is good, though, it would be nice if you could replace your 'tab indents' with 'two whitespace indents' instead, just so that all the code looks the same. I will need to get some more people to test it before I merge it in, just to be sure that we don't breaky anything.
Thanks for your contribution!

icc commented on admin/class-h5p-plugin-admin.php:
```diff
@@ -646,7 +664,7 @@ private static function asset_handle($path) {
    */
   public static function add_script($handle, $path) {
-    wp_enqueue_script(self::asset_handle($handle), plugins_url('h5p/' . $path), array(), H5P_Plugin::VERSION);
+    wp_enqueue_script(self::asset_handle($handle), plugins_url('h5p/' . $path), array(), H5P_Plugin::VERSION, true);
```
Any reason we need to change it to always be in the footer? I'm a little worried that this may break something in some themes.

rcarcasses replied:
Oh, I think this is not necessary at all, sorry for that. Is there any easy way of rejecting this from your side or do I have to update the pull request (sorry for my ignorance)?

rcarcasses commented:
No problem at all, I'll check the idents problem and will update the pull request soon. Thank you!

rcarcasses commented:
Hi @icc I have updated the code:

- Identation now uses two spaces convention.
- Removed the non-relevant line of code related to the wp_enqueue_script function.

We are building an application which uses as a fundamental block h5p: we have created an exerciser which gives a sequence of h5p exercises and this sequence is not linear in the sense that the next h5p depends on the response of the users at the current one or in some cases the user is presented with several options which she/he can choose to do. Generally speaking is a tree of h5p in which the exercise start at the root h5p and ends whenever the user reach one of the leaves. We believe the strength of our tool will be the ability of designing custom paths in this tree depending on user's input. 

So that's why we need a custom way of displaying the h5p editor, we don't want end-users to go to the wordpress admin to create new content but to given them the ability to create several and to choose how them will be connected.

We are close to releasing it and we will be happy to share it with you of course!

Once more thank you,
Robert

icc commented:
That is great to hear and thanks for sharing your use case. I will make sure to get this tested and merged in as soon as possible, thank you!
